src/tribler-core/tribler_core/modules/community_loader.py

```diff
@@ -0,0 +1,141 @@
+
+
+def dht_discovery_community():
+    from ipv8.dht.discovery import DHTDiscoveryCommunity
+    return DHTDiscoveryCommunity
+
+
+def random_churn():
+    from ipv8.peerdiscovery.churn import RandomChurn
+    return RandomChurn
+
+
+def ping_churn():
+    from ipv8.dht.churn import PingChurn
+    return PingChurn
+
+
+def random_walk():
+    from ipv8.peerdiscovery.discovery import RandomWalk
+    return RandomWalk
+
+
+def periodic_similarity():
+    from ipv8.peerdiscovery.community import PeriodicSimilarity
+    return PeriodicSimilarity
+
+
+@precondition('session.config.discovery_community.enabled')
+@overlay(discovery_community)
+@kwargs(max_peers='100')
+@walk_strategy(random_churn, target_peers=INFINITE)
+@walk_strategy(random_walk)
+@walk_strategy(periodic_similarity, target_peers=INFINITE)
+class IPv8DiscoveryCommunityLauncher(TriblerCommunityLauncher):
+    pass
+
+
+@precondition('session.config.dht.enabled')
+@set_in_session('dht_community')
+@overlay(dht_discovery_community)
+@kwargs(max_peers='60')
+@walk_strategy(ping_churn, target_peers=INFINITE)
+@walk_strategy(random_walk)
+class DHTCommunityLauncher(TriblerCommunityLauncher):
+    pass
+
+
+def create_default_loader(config: TriblerConfig, tunnel_testnet: bool = False,
+                          bandwidth_testnet: bool = False, chant_testnet: bool = False):
+    loader = IPv8CommunityLoader()
+
+    loader.set_launcher(IPv8DiscoveryCommunityLauncher())
+    loader.set_launcher(DHTCommunityLauncher())
+
+    if bandwidth_testnet:
+        from tribler_core.modules.bandwidth_accounting.launcher import BandwidthTestnetCommunityLauncher
+        loader.set_launcher(BandwidthTestnetCommunityLauncher())
+    else:
+        from tribler_core.modules.bandwidth_accounting.launcher import BandwidthCommunityLauncher
+        loader.set_launcher(BandwidthCommunityLauncher())
+
+    if config.tunnel_community.enabled and not tunnel_testnet:
+        from tribler_core.modules.tunnel.community.launcher import TriblerTunnelCommunityLauncher
+        loader.set_launcher(TriblerTunnelCommunityLauncher())
+
+    if config.tunnel_community.enabled and tunnel_testnet:
+        from tribler_core.modules.tunnel.community.launcher import TriblerTunnelTestnetCommunityLauncher
+        loader.set_launcher(TriblerTunnelTestnetCommunityLauncher())
+
+    if config.popularity_community.enabled:
+        from tribler_core.modules.popularity.launcher import PopularityCommunityLauncher
+        loader.set_launcher(PopularityCommunityLauncher())
+
+    if config.chant.enabled and not chant_testnet:
+        from tribler_core.modules.metadata_store.community.launcher import GigaChannelCommunityLauncher
+        loader.set_launcher(GigaChannelCommunityLauncher())
+
+    if config.chant.enabled and chant_testnet:
+        from tribler_core.modules.metadata_store.community.launcher import GigaChannelTestnetCommunityLauncher
+        loader.set_launcher(GigaChannelTestnetCommunityLauncher())
```

Comment:
I have two problems with this approach:

 1. It extracts the internal logic of a launcher (a module) outside of its scope. **Launchers are no longer self-contained.**
 2. This has now turned from managing abstract launchers, to actively setting up the launcher internals (checking launch conditions). **This file has turned into a god object.**

I agree that this makes it easier to work with an IDE, allowing top-level imports in the `launcher.py` definitions. However, ultimately, I don't think IDE integration for files we touch once every +- year is worth the **sacrifice of demodularizing the Tribler architecture**. Therefore, I consider this part of your PR an unacceptable regression in architectural integrity.

I'd be happy to help with a solution that allows top-level imports without the problems I stated above. For now, I'd suggest reverting these lines.

Reply:
Thank you for this comment. Will try to explain my reasoning.

> It extracts the internal logic of a launcher (a module) outside of its scope

It extracts conditions from a launcher to a loader. Yes. 
But I would disagree with the statement that conditions are an internal and integral part of the launcher. Conditions describe the high-level composition of launchers and then communities. It looks more natural to set this composition of launchers explicitly.
Launchers become more self-contained in this case. 

If I need a custom composition of launchers I will not edit the launchers themselves, but just define a different loader.
https://github.com/drew2a/tribler/blob/ada438d691d1e14b3da743f0c2f6e7a6cc8fa7fb/src/tribler-core/run_bandwidth_crawler.py#L40-L47

> This file has turned into a god object.

Yes, the file has turned to something different, but I doubt it is a god object now. 
It actually was a "god" not object, but file before. 
This single file consisted of launchers for all communities and loaders for all communities. 

Now `community_loader.py` responds only for a default set of launchers. (It contains also `TriblerCommunityLauncher` and `TestnetMixIn`, and yes we can extract them into separate files too)

>However, ultimately, I don't think IDE integration for files we touch once every +- year is worth the sacrifice of demodularizing the Tribler architecture. 

I would gladly agree with you, but I don't see "demodularisation" here. Moreover, I don't see modularization in Tribler. It is a kind of "some degree" of modularization. But the logic of modules, however highly intertwined and not isolated.

Below just one single example:
https://github.com/Tribler/tribler/blob/814a94ad5faaa915083c9c9b299a831684e28802/src/tribler-core/tribler_core/session.py#L306-L316

Short description: if chant enables, then create `mds`-instance, which is used in many places:
![image](https://user-images.githubusercontent.com/13798583/123510372-5d536500-d67b-11eb-85ba-146a5bfb6571.png)

So, from my perspective, this PR is a step closer to modularisation. 
It extracts loader logic (that is an integral part of a community)  from one single big file next to the community itself.

Reply:
> Conditions describe the high-level composition of launchers and then communities. It looks more natural to set this composition of launchers explicitly.
> Launchers become more self-contained in this case.
>
> If I need a custom composition of launchers I will not edit the launchers themselves, but just define a different loader.

Why does it look more natural to you?
This does not fit a module pattern at all.
This fits a mediator pattern.

A module should initialize and finalize itself.
In your approach the manager of launchers (i.e., the loader) actively inspects and checks launcher-specific logic.
This also means that to add a launcher you need to add a loader.
To give an example analogy, this is like recompiling your operating system every time you install a new application.
I don't think this is natural at all.

> Yes, the file has turned to something different, but I doubt it is a god object now.
> It actually was a "god" not object, but file before.
> This single file consisted of launchers for all communities and loaders for all communities.

You have converted a "manager of launchers" into "a list of if-statements that deal with the internal logic of launchers".
This kills the abstraction of modules.
God file or god object, whatever you want to call it.

Still, though, I am not against moving each launcher to a separate file.
I do believe that is indeed the next step in modularization.

> [...] I don't see "demodularisation" here. Moreover, I don't see modularization in Tribler. It is a kind of "some degree" of modularization. But the logic of modules, however highly intertwined and not isolated.

The demodularization is in the breaking of the module design pattern I discussed before, I won't repeat it again.
Two further points on this though:

 1. Modules can be intertwined, that doesn't make them less modular. Just like `pip` dependencies have dependencies themselves.
 2. Just because Tribler modularization is not complete yet, it doesn't mean we should just void all previous work towards it. An analogy: "Your child can walk, but it can't ride a bike. So, we're breaking its legs".

> It extracts loader logic (that is an integral part of a community) from one single big file next to the community itself.

A single file can specify modules, that doesn't make the modules less modular.
At the same time, you can split highly centralized logic into multiple files just the same.
I don't see how the amount of files relates to the degree of modularity.
That said (repeating), I still see the value in placing the individual launcher files next to the module directories.

Reply:
> A module should initialize and finalize itself.
In your approach the manager of launchers (i.e., the loader) actively inspects and checks launcher-specific logic.
This also means that to add a launcher you need to add a loader.

Not sure. In my approach loader doesn't check any launcher-specific logic actually. 
In my approach, it specifies the composition of launchers (which can be unique for any use case of `Session`). 

If you like to use analogies it is like to specify `requirements.txt` for each specific application instead of using one single big complex `requirements.txt` that contains all possible modules and their conditions.

> You have converted a "manager of launchers" into "a list of if-statements that deal with the internal logic of launchers".

I would say, I've converted the file, that was responsible for:
1. Defining launchers to each community
1. Defining default set of launchers

into:

1. Four launchers located next to corresponding communities
1. Default loader that used in Tribler main application (Tribler itself)

That said I have split the logic of something named "module catalog"(which actually had nothing to do with modules, but managed the loading of communities instead) into `community_loader.py` and `launcher.py`  

> Just because Tribler modularization is not complete yet, it doesn't mean we should just void all previous work towards it. An analogy: "Your child can walk, but it can't ride a bike. So, we're breaking its legs".

First of all, emphasize that, `ipv8_module_catalog` is not a part of the Tribler "modules", but a part of Tribler communities, and in particular, it is a loader manager for communities. Let's call things by their proper names.

Next, again, If you like analogies, then "I'm trying to give your baby a bike, to give him a chance to learn how to ride"

Reply:
> Not sure. In my approach loader doesn't check any launcher-specific logic actually.

`config.tunnel_community.enabled and not tunnel_testnet` checks two configuration options of the TunnelCommunity outside of the TunnelCommunity module.

`config.chant.enabled and not chant_testnet` checks two configuration options of the GigaChannelCommunity outside of the GigaChannelCommunity module.

etc.

This is very much module specific. There is no abstraction from the component internals.

> In my approach, it specifies the composition of launchers (which can be unique for any use case of Session).

Doing this outside of a module itself is a textbook mediator pattern.

> If you like to use analogies it is like to specify requirements.txt for each specific application instead of using one single big complex requirements.txt that contains all possible modules and their conditions.

No, you're doing exactly the opposite. You're taking the individual requirements and pulling them into a long list. This is why I request changes.

> > You have converted a "manager of launchers" into "a list of if-statements that deal with the internal logic of launchers".
>
> I would say, I've converted the file, that was responsible for:
>
> 1. Defining launchers to each community
> 2. Defining default set of launchers
> 
> into:
>
> 1. Four launchers located next to corresponding communities
> 2. Default loader that used in Tribler main application (Tribler itself)

I agree and I like the idea. However, one does not exclude the other. Through your changes you changed the responsiblities of loaders from simple managers to managers with embedded elements (see [my other answer](https://github.com/Tribler/tribler/pull/6182/files/ada438d691d1e14b3da743f0c2f6e7a6cc8fa7fb#r659699915) -- I hope that clarifies this more).

> That said I have split the logic of something named "module catalog"(which actually had nothing to do with modules, but managed the loading of communities instead) into community_loader.py and launcher.py

It has everything to do with modules. These launchers are simply elaborate configuration definitions, not the actual modules themselves. I believe you misunderstand the nature of launchers and the loader: modules are not actually loaded here (see also [my other answer](https://github.com/Tribler/tribler/pull/6182/files/ada438d691d1e14b3da743f0c2f6e7a6cc8fa7fb#r659699915)).

> Next, again, If you like analogies, then "I'm trying to give the baby a bike, to give him a chance to learn how to ride"

I like analogies and I think this is noble goal. I also like moving files next to modules. My analogy is what I inferred from your code and it does not match your analogy. This is why I object. I object because I do not see your goal align with your implementation.

Reply:
> config.tunnel_community.enabled and not tunnel_testnet checks two configuration options of the TunnelCommunity outside of the TunnelCommunity module.
config.chant.enabled and not chant_testnet checks two configuration options of the GigaChannelCommunity outside of the GigaChannelCommunity module.
etc.
This is very much module specific. There is no abstraction from the component internals.

Yes, you can look at them in this way. 

From my POV in your approach, launchers (by defining loading conditions that you described above) have an impact on the higher structure — composition of launchers. From my POV it is wrong. They should not care about their composition. But they can care about the order of load (you nicely implemented it in `after` annotation). 
(actually, the load order is another very discussable topic, but let's skip it for this PR).

I want a more flexible and explicit approach in loaders. 
I feel a lack of explicitly and of flexibility here. 

You can get it as feedback from a developer who tried to use the current system and was disappointed (many-many times). As I see it from code I'm not alone. 

But besides just the feedback, I spent time developing an option to present how the current system can be improved. 
You see it is as a degradation (that's fine). I see it is as an improvement.

Make note that this PR fits your limitations on dynamic loading (which you patiently explained to me during the consultation I took before I started PR).

Let's try to make an arrangement for this PR tomorrow. If we can't find a middle ground, then will take a second opinion from a team :) 🤝 

>Doing this outside of a module itself is a textbook mediator pattern.

I agree it is kind-a-mediator, but the mediator "encapsulates the interaction between a set of objects" and our loader just loads a set (or a composition) of communities via launchers.